src/tooling/docs-assembler/Sourcing/RepositorySourcesFetcher.cs

```diff
@@ -46,129 +47,169 @@ public IReadOnlyCollection<Checkout> GetAll()
+			{
+				_logger.LogError(e, "{RepositoryName}: Failed to acquire current commit, falling back to recreating from scratch", repository.Name);
+				checkoutFolder.Delete(true);
+				checkoutFolder.Refresh();
+				return CloneRef(repository, gitRef, pull, attempt + 1);
+			}
 		}
-		catch (Exception e)
+		// Repository already checked out the same commit
+		if (head != null && head == gitRef)
+			// nothing to do, already at the right commit
+			_logger.LogInformation("{RepositoryName}: HEAD already at {GitRef}", repository.Name, gitRef);
+		else
 		{
-			_logger.LogError(e, "Failed to update {Name} from {RelativePath}, falling back to recreating from scratch", name, relativePath);
-			if (checkoutFolder.Exists)
+			FetchAndCheckout(repository, gitRef, checkoutFolder);
+			if (!pull)
+			{
+				return new Checkout
+				{
+					Directory = checkoutFolder,
+					HeadReference = gitRef,
+					Repository = repository,
+				};
+			}
+			try
+			{
+				ExecIn(checkoutFolder, "git", "pull", "--depth", "1", "--allow-unrelated-histories", "--no-ff", "origin", gitRef);
+			}
+			catch (Exception e)
 			{
+				_logger.LogError(e, "{RepositoryName}: Failed to update {GitRef} from {RelativePath}, falling back to recreating from scratch",
+					repository.Name, gitRef, checkoutFolder.FullName);
 				checkoutFolder.Delete(true);
 				checkoutFolder.Refresh();
+				return CloneRef(repository, gitRef, pull, attempt + 1);
 			}
-			return false;
 		}
 
-		head = Capture(checkoutFolder, "git", "rev-parse", "HEAD");
+		return new Checkout
+		{
+			Directory = checkoutFolder,
+			HeadReference = gitRef,
+			Repository = repository,
+		};
+	}
 
-		return true;
+	/// <summary>
+	/// Initializes the git repository if it is not already initialized.
+	/// Returns true if the repository was already initialized.
+	/// </summary>
+	private bool GitInit(Repository repository, IDirectoryInfo checkoutFolder)
+	{
+		var isGitAlreadyInitialized = Directory.Exists(Path.Combine(checkoutFolder.FullName, ".git"));
+		if (isGitAlreadyInitialized)
+			return true;
+		ExecIn(checkoutFolder, "git", "init");
+		ExecIn(checkoutFolder, "git", "remote", "add", "origin", repository.Origin);
+		return false;
 	}
 
-	private string CheckoutFromScratch(Repository repository, string name, string branch, string relativePath, IDirectoryInfo checkoutFolder)
+	private void FetchAndCheckout(Repository repository, string gitRef, IDirectoryInfo checkoutFolder)
 	{
-		_logger.LogInformation("Checkout: {Name}\t{Branch}\t{RelativePath}", name, branch, relativePath);
+		ExecIn(checkoutFolder, "git", "fetch", "--no-tags", "--prune", "--no-recurse-submodules", "--depth", "1", "origin", gitRef);
 		switch (repository.CheckoutStrategy)
 		{
-			case "full":
-				Exec("git", "clone", repository.Origin, checkoutFolder.FullName,
-					"--depth", "1", "--single-branch",
-					"--branch", branch
-				);
+			case CheckoutStrategy.Full:
+				ExecIn(checkoutFolder, "git", "sparse-checkout", "disable");
```

Comment:
Why is this necessary? (not questioning it just looking to learn).

Reply:
It's actually unlikely.. But I was just thinking, if you ever change the checkout strategy and if you run the command, while the repo is still checked out (cached) you want to also disabled it.